Import state doesn't work with Bor consensus

**Error details**

Since Bor uses `consensus/bor.go` consensus, it is failing while fetching the validators

```
INFO [03-15|23:09:37.578] Bumping default cache on mainnet         provided=1024 updated=4096
INFO [03-15|23:09:37.581] Maximum peer count                       ETH=50 LES=0 total=50
INFO [03-15|23:09:37.607] Allocated cache and file handles         database=/Users/jdkanani/matic/testnets/bor-devnet/node1/bor/chaindata cache=2.00GiB handles=5120
INFO [03-15|23:09:37.695] Opened ancient database                  database=/Users/jdkanani/matic/testnets/bor-devnet/node1/bor/chaindata/ancient
INFO [03-15|23:09:38.040] Loaded most recent local header          number=0 hash=13b5e3…d73f00 td=1 age=10mo7h7m
INFO [03-15|23:09:38.040] Loaded most recent local full block      number=0 hash=13b5e3…d73f00 td=1 age=10mo7h7m
INFO [03-15|23:09:38.040] Loaded most recent local fast block      number=0 hash=13b5e3…d73f00 td=1 age=10mo7h7m
INFO [03-15|23:09:38.040] Importing blockchain                     file=state.json
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x4705b40]

goroutine 68 [running]:
github.com/maticnetwork/bor/internal/ethapi.(*PublicBlockChainAPI).Call(0x0, 0x5028e00, 0xc0000dc600, 0x0, 0xc0001d3540, 0xc00017a0c0, 0x0, 0x0, 0xc000176400, 0x0, ...)
	/Users/jdkanani/go/src/github.com/maticnetwork/bor/build/_workspace/src/github.com/maticnetwork/bor/internal/ethapi/api.go:815 +0x50
github.com/maticnetwork/bor/consensus/bor.(*Bor).GetCurrentValidators(0xc0000eec60, 0x0, 0x1, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/jdkanani/go/src/github.com/maticnetwork/bor/build/_workspace/src/github.com/maticnetwork/bor/consensus/bor/bor.go:950 +0x4cb
github.com/maticnetwork/bor/consensus/bor.(*Bor).snapshot(0xc0000eec60, 0x502f880, 0xc04259e000, 0x0, 0x6b9ccba97fe3b513, 0x5870f002ceeed546, 0x1da4cb9ecc22c5f8, 0x3fd7582f6f6e08, 0xc01d456c30, 0x0, ...)
	/Users/jdkanani/go/src/github.com/maticnetwork/bor/build/_workspace/src/github.com/maticnetwork/bor/consensus/bor/bor.go:469 +0x4ea
github.com/maticnetwork/bor/consensus/bor.(*Bor).verifyCascadingFields(0xc0000eec60, 0x502f880, 0xc04259e000, 0xc0001a7680, 0xc01d456c30, 0x0, 0x5, 0x0, 0x0)
	/Users/jdkanani/go/src/github.com/maticnetwork/bor/build/_workspace/src/github.com/maticnetwork/bor/consensus/bor/bor.go:403 +0x151
github.com/maticnetwork/bor/consensus/bor.(*Bor).verifyHeader(0xc0000eec60, 0x502f880, 0xc04259e000, 0xc0001a7680, 0xc01d456c30, 0x0, 0x5, 0x0, 0x0)
	/Users/jdkanani/go/src/github.com/maticnetwork/bor/build/_workspace/src/github.com/maticnetwork/bor/consensus/bor/bor.go:372 +0x288
github.com/maticnetwork/bor/consensus/bor.(*Bor).VerifyHeaders.func1(0xc01d456c30, 0x5, 0x5, 0xc0000eec60, 0x502f880, 0xc04259e000, 0xc000406420, 0xc000105020)
	/Users/jdkanani/go/src/github.com/maticnetwork/bor/build/_workspace/src/github.com/maticnetwork/bor/consensus/bor/bor.go:308 +0xa5
created by github.com/maticnetwork/bor/consensus/bor.(*Bor).VerifyHeaders
	/Users/jdkanani/go/src/github.com/maticnetwork/bor/build/_workspace/src/github.com/maticnetwork/bor/consensus/bor/bor.go:306 +0xda
```

**Reproduction steps**

Export state

```
bor --datadir /Users/jdkanani/matic/testnets/bor-devnet/node1 export state.json
```

Import state in clean bor setup

```
bor --datadir /Users/jdkanani/matic/testnets/bor-devnet/node1 import state.json
```

**Version**

Branch: `master`

--- a/cmd/bor/chaincmd.go
+++ b/cmd/bor/chaincmd.go
@@ -61,7 +61,7 @@ It expects the genesis file as argument.`,
 		Action:    utils.MigrateFlags(importChain),
 		Name:      "import",
 		Usage:     "Import a blockchain file",
-		ArgsUsage: "<filename> (<filename 2> ... <filename N>) ",
+		ArgsUsage: "<filename> (<filename 2> ... <filename N>) <genesisPath>",
 		Flags: []cli.Flag{
 			utils.DataDirFlag,
 			utils.CacheFlag,
@@ -228,8 +228,8 @@ func initGenesis(ctx *cli.Context) error {
 }
 
 func importChain(ctx *cli.Context) error {
-	if len(ctx.Args()) < 1 {
-		utils.Fatalf("This command requires an argument.")
+	if len(ctx.Args()) < 2 {
+		utils.Fatalf("This command requires atleast 2 arguments.")
 	}
 	stack := makeFullNode(ctx)
 	defer stack.Close()
@@ -255,12 +255,13 @@ func importChain(ctx *cli.Context) error {
 	// Import the chain
 	start := time.Now()
 
-	if len(ctx.Args()) == 1 {
+	// ArgsUsage: "<filename> (<filename 2> ... <filename N>) <genesisPath>",
+	if len(ctx.Args()) == 2 {
 		if err := utils.ImportChain(chain, ctx.Args().First()); err != nil {
 			log.Error("Import error", "err", err)
 		}
 	} else {
-		for _, arg := range ctx.Args() {
+		for _, arg := range ctx.Args()[:len(ctx.Args())-1] {
 			if err := utils.ImportChain(chain, arg); err != nil {
 				log.Error("Import error", "file", arg, "err", err)
 			}

--- a/cmd/utils/flags.go
+++ b/cmd/utils/flags.go
@@ -1661,11 +1661,8 @@ func MakeGenesis(ctx *cli.Context) *core.Genesis {
 	return genesis
 }
 
-func getGenesis(ctx *cli.Context) *core.Genesis {
-	genesisPath := ctx.Args().Get(len(ctx.Args()) - 1)
-	if genesisPath == "" {
-		return nil
-	}
+func getGenesis(genesisPath string) *core.Genesis {
+	log.Info("Reading genesis at ", "file", genesisPath)
 	file, _ := os.Open(genesisPath)
 	defer file.Close()
 
@@ -1676,10 +1673,8 @@ func getGenesis(ctx *cli.Context) *core.Genesis {
 
 // MakeChain creates a chain manager from set command line flags.
 func MakeChain(ctx *cli.Context, stack *node.Node) (chain *core.BlockChain, chainDb ethdb.Database) {
-	genesis := getGenesis(ctx)
-	if genesis == nil {
-		genesis = MakeGenesis(ctx)
-	}
+	// expecting the last argument to be the genesis file
+	genesis := getGenesis(ctx.Args().Get(len(ctx.Args()) - 1))
 
 	var err error
 	chainDb = MakeChainDatabase(ctx, stack)

--- a/cmd/utils/flags.go
+++ b/cmd/utils/flags.go
@@ -19,6 +19,7 @@ package utils
 
 import (
 	"crypto/ecdsa"
+	"encoding/json"
 	"errors"
 	"fmt"
 	"io/ioutil"
@@ -34,6 +35,7 @@ import (
 	"github.com/maticnetwork/bor/common"
 	"github.com/maticnetwork/bor/common/fdlimit"
 	"github.com/maticnetwork/bor/consensus"
+
 	"github.com/maticnetwork/bor/consensus/bor"
 	"github.com/maticnetwork/bor/consensus/clique"
 	"github.com/maticnetwork/bor/consensus/ethash"
@@ -47,6 +49,8 @@ import (
 	"github.com/maticnetwork/bor/ethdb"
 	"github.com/maticnetwork/bor/ethstats"
 	"github.com/maticnetwork/bor/graphql"
+
+	pcsclite "github.com/gballet/go-libpcsclite"
 	"github.com/maticnetwork/bor/les"
 	"github.com/maticnetwork/bor/log"
 	"github.com/maticnetwork/bor/metrics"
@@ -61,7 +65,6 @@ import (
 	"github.com/maticnetwork/bor/params"
 	"github.com/maticnetwork/bor/rpc"
 	whisper "github.com/maticnetwork/bor/whisper/whisperv6"
-	pcsclite "github.com/gballet/go-libpcsclite"
 	cli "gopkg.in/urfave/cli.v1"
 )
 
@@ -1658,19 +1661,68 @@ func MakeGenesis(ctx *cli.Context) *core.Genesis {
 	return genesis
 }
 
+func getGenesis(ctx *cli.Context) *core.Genesis {
+	genesisPath := ctx.Args().Get(len(ctx.Args()) - 1)
+	if genesisPath == "" {
+		return nil
+	}
+	file, _ := os.Open(genesisPath)
+	defer file.Close()
+
+	genesis := new(core.Genesis)
+	json.NewDecoder(file).Decode(genesis)
+	return genesis
+}
+
 // MakeChain creates a chain manager from set command line flags.
 func MakeChain(ctx *cli.Context, stack *node.Node) (chain *core.BlockChain, chainDb ethdb.Database) {
+	genesis := getGenesis(ctx)
+	if genesis == nil {
+		genesis = MakeGenesis(ctx)
+	}
+
 	var err error
 	chainDb = MakeChainDatabase(ctx, stack)
-	config, _, err := core.SetupGenesisBlock(chainDb, MakeGenesis(ctx))
+	config, _, err := core.SetupGenesisBlock(chainDb, genesis)
 	if err != nil {
 		Fatalf("%v", err)
 	}
+
 	var engine consensus.Engine
+	var ethereum *eth.Ethereum
 	if config.Clique != nil {
 		engine = clique.New(config.Clique, chainDb)
 	} else if config.Bor != nil {
-		engine = bor.New(config, chainDb, nil)
+		cfg := &eth.Config{Genesis: genesis}
+		workspace, err := ioutil.TempDir("", "console-tester-")
+		if err != nil {
+			Fatalf("failed to create temporary keystore: %v", err)
+		}
+
+		// Create a networkless protocol stack and start an Ethereum service within
+		stack, err := node.New(&node.Config{DataDir: workspace, UseLightweightKDF: true, Name: "console-tester"})
+		if err != nil {
+			Fatalf("failed to create node: %v", err)
+		}
+		err = stack.Register(func(ctx *node.ServiceContext) (node.Service, error) {
+			s, err := eth.New(ctx, cfg)
+			return s, err
+		})
+		if err != nil {
+			Fatalf("failed to register Ethereum protocol: %v", err)
+		}
+
+		// Start the node and assemble the JavaScript console around it
+		if err = stack.Start(); err != nil {
+			Fatalf("failed to start test stack: %v", err)
+		}
+		_, err = stack.Attach()
+		if err != nil {
+			Fatalf("failed to attach to node: %v", err)
+		}
+
+		stack.Service(&ethereum)
+		engine = ethereum.Engine().(*bor.Bor)
 	} else {
 		engine = ethash.NewFaker()
 		if !ctx.GlobalBool(FakePoWFlag.Name) {
@@ -1702,6 +1754,9 @@ func MakeChain(ctx *cli.Context, stack *node.Node) (chain *core.BlockChain, chai
 	}
 	vmcfg := vm.Config{EnablePreimageRecording: ctx.GlobalBool(VMEnableDebugFlag.Name)}
 	chain, err = core.NewBlockChain(chainDb, cache, config, engine, vmcfg, nil)
+	if ethereum != nil {
+		ethereum.SetBlockchain(chain)
+	}
 	if err != nil {
 		Fatalf("Can't create BlockChain: %v", err)
 	}

--- a/eth/backend.go
+++ b/eth/backend.go
@@ -231,6 +231,10 @@ func New(ctx *node.ServiceContext, config *Config) (*Ethereum, error) {
 	return eth, nil
 }
 
+func (s *Ethereum) SetBlockchain(blockchain *core.BlockChain) {
+	s.blockchain = blockchain
+}
+
 func makeExtraData(extra []byte) []byte {
 	if len(extra) == 0 {
 		// create default extradata

